JDBC sample app to show use of Auth Tokens to connect the database and get the tokens from endpoints

rvpanchal1202 opened the pull request:
-> Review Comments are addressed
-> Testing of functionality is done with Keyclock IDP

rvpanchal1202 committed:
Update OAuthSampleApp.java, Added refresh token api to get new access token

using the refresh token, we can get the new access token in case of a wrong access token or an expired access token. 

Added API to get the new access_token using refresh_token

note: When we request for new access_token using refresh token, This request will give us both the tokens. which we can save later to original tokens and start using new Tokens instead of old tokens

rvpanchal1202 committed:
Update OAuthSampleApp.java

Changes in the client application done as per comments and discussion. 

SetUpDbForOAuth() -> Adding authentication record in Vertica database is having issues. It's dropping the user and authentication method, but while setting the Authentication record, there is some error because of which Token authentication is failing and Database connection with access token is unable to establish. 

Without SetUpDbForOAuth() method call, Application is woking good.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Commented SetUpDbForOAuth() and TearDown() call.

rvpanchal1202 committed:
Update OAuthSampleApp.java

-> Removed all @commandline and picoli imports
-> Removed all the blank lines from the file
-> Indentation issue resolved. Note:Please adjust the tab stop as 4 to view the file content properly indented.
-> Removed if else for first element in the map for framing the connectionString. 
-> All the Start braces and end braces are aligned and made the same in every place 
-> Added null check for an access token and refresh token after getting it from token URL

rvpanchal1202 committed:
Create config.properties

Added config.properties file which would be required for running the sample app. User have to update the values as per configuration. 

Note: Given Sample example for the Connection string. Because Connection string will be accepted in this format. <KeyName>=<Value>;<KeyName>=<Value>;......

rvpanchal1202 committed:
Update OAuthSampleApp.java

-> Added functions for the redundant code for copying data to the buffered output stream.
-> removed unnecessary imports.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Removed ClientID, ClientSecret, Tokenurl from getconnection Call.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Added Line between imports and class

rvpanchal1202 committed:
Update OAuthSampleApp.java

Debug prints removed

rvpanchal1202 committed:
Update OAuthSampleApp.java

Deprecated method use JsonParser().parse(res); removed. 
used JsonParser.parseString(res);

rvpanchal1202 committed:
Update OAuthSampleApp.java

- Added Fix for setup of Authentication record to Database. 
- Uncommented DBSetup and Teardown for Authentication setup before OAuth sample test run

rvpanchal1202 committed:
Update config.properties

Added DiscoveryUrl to be provided in config.properties file for OAuth Authentication

rvpanchal1202 committed:
Update OAuthSampleApp.java

Added RefreshToken check and If both the tokens are available, then it will use those tokens, Or else it will get new tokens using password grant.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Check if access token is not available, Get the tokens.
else 
if a refresh token is not available, Get the tokens.
else 
set both the tokens to properties.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Indentation issue resolved

rvpanchal1202 committed:
Update OAuthSampleApp.java

Indentation resolved

rvpanchal1202 committed:
Update OAuthSampleApp.java

Comment indentation resolved

rvpanchal1202 committed:
Update OAuthSampleApp.java

Internal review comments are addressed. 
-> Redundant code for getting parameters from connection string addressed.
-> Formating of comments for method explanation addressed.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Review comments are addressed. 
Changes ensureAccessToken as per suggestion given
Comments and print statements are added

rvpanchal1202 committed:
Update OAuthSampleApp.java

-> Spacing between Functions and indentation formating is done

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,298 @@
+	private static void SetUpDbForOAuth() throws Exception {
+		String connectionString = getConnectionString();
+		vConnection = DriverManager.getConnection(connectionString);
+		String CONNECTION_STRING = prop.getProperty("ConnectionString");
```
Is CONNECTION_STRING still needed for something? It's not used anywhere.

rvpanchal1202 replied:
No, removed the variable CONNECTION_STRING, Its unused variable in this function after adding other calls for getting connection String.

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,298 @@
+					break;
+				}
+				try {
+					GetTokensByRefreshGrant();
```
I'd put a message here saying that the access token is invalid/expired, trying to do refresh.

rvpanchal1202 replied:
-> added a message

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,298 @@
+					GetTokensByRefreshGrant();
+				} catch (Exception e1) {
+					try {
+						GetTokensByPasswordGrant();
```
Another message here would be good saying refresh token is invalid/expired, getting new tokens.

rvpanchal1202 replied:
Added a message

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,298 @@
+	}
+
+	// If there is no access token, Get it using password grant
+	private static void EnsureAccessToken() throws Exception {
```
The logic to check for a refresh token if there is no access token is not in this function.

rvpanchal1202 replied:
Added the logic to check for a refresh token if there is no access Token

rvpanchal1202 committed:
Update OAuthSampleApp.java

- Unused variable CONNECTION_STRING removed. 
- Added comments for the required calls
- Added Refresh token check code in EnsureAccessToken Call

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,310 @@
+					try{
+						System.out.println("Attempting to use given refresh token.");
+						GetTokensByRefreshGrant();
+					}catch(){
```
This catch is empty

rvpanchal1202 replied:
Added exception variable for catch block

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,310 @@
+			SetUpDbForOAuth();
+			EnsureAccessToken();
+			ConnectToDatabase();
+		} catch (SQLTransientConnectionException connException) {
```
Do you think it's necessary to catch these three SQL exceptions before the line on 312? I'm not sure, it's up to you.

rvpanchal1202 replied:
Yes, We can have a single exception handle for all these SQLExceotions. It was already written all these exception in old app. Removed now.

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,310 @@
+				// refresh_token
+				String accessToken = jObject.has("access_token") ? jObject.get("access_token").getAsString() : null;
+				String refreshToken = jObject.has("refresh_token") ? jObject.get("refresh_token").getAsString() : null;
+				if (null == accessToken || null == refreshToken) {
```
Is this is a little off. I said that sometimes there is no refresh token but there is still an access token. This would throw an exception even if we get a good access token.

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,310 @@
+		}
+	}
+
+	// If there is no access token, Get it using password grant
```
This comment needs to be updated.

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,310 @@
+	}
+
+	// Connect to Database using access Token
+	private static Connection connectToDB(String accessToken) throws SQLException {
```
You have a function called "connectToDB" and one called "ConnectToDatabase". This would be confusing for users.

rvpanchal1202 replied:
Changed function name from connecToDb to getConnection.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Refresh Token check is removed as it might be that we will get only access Token and not refresh token. As per comment from Kevin.

rvpanchal1202 committed:
Update OAuthSampleApp.java

-> Updated comment for EnsureAccessToken

rvpanchal1202 committed:
Update OAuthSampleApp.java

Changed the function name from connectToDb to getConnection.

kevinkarch88 commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,307 @@
+							"Access/refresh token is null, Please verify the config.properties for proper inputs.");
+				}
+				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, accessToken);
+				System.setProperty(OAUTH_REFRESH_TOKEN_VAR_STRING, refreshToken);
```
What happens if the refresh token is null? Maybe we need a null check here.

rvpanchal1202 replied:
It will assign null or empty string to refreshToken.

kevinkarch88 replied:
I think storing an environment variable as 'null' is a bad idea.

rvpanchal1202 replied:
**Added the null and empty check condition to check the refresh token, and store only if its not null/not empty.**

rvpanchal1202 committed:
Update OAuthSampleApp.java

Added null and empty checks for refresh token to save in the properties.

kevinkarch88 commented:
I don't have any major issues with the code as is, it should work as intended.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+public class OAuthSampleApp {
+	private static Properties prop;
+	private static Map<String, String> csProp = new HashMap<String, String>();
+	private static String OAUTH_ACCESS_TOKEN_VAR_STRING = "OAUTH_SAMPLE_ACCESS_TOKEN";
```
In java we declare constants as `final` so the constant properties here should be `private static final`

Constants are also generally grouped together and declared first in a class, so I would move these two constants to be the first properties defined in the class followed by a newline and then followed by the rest of the properties

rvpanchal1202 replied:
Changed OAUTH_ACCESS_TOKEN_VAR_STRING, OAUTH_REFRESH_TOKEN_VAR_STRING to constant "private static final"

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+import com.google.gson.JsonObject;
+import com.vertica.jdbc.VerticaConnection;
+
+public class OAuthSampleApp {
```
Unless it is just rendering poorly in github, the spacing in this class is all off. Indentation should simply be four spaces per indentation level and they should be actual spaces, not tabs.

rvpanchal1202 replied:
Hi Danny, 
This indentation formatting is done using the auto-indent feature(Source->Format) in Eclipse. So If we make any changes here in git for indentation, It will reflect wrong when we open the file in Eclipse or other IDE.

DMickens replied:
If you are going to use the auto indent feature you need to update it to use four spaces instead of tabs, there should be an option for this. A space is a space, it's the same everywhere. A tab character may not render the same everywhere like we see here

rvpanchal1202 replied:
Ok. As per the last commit, All indentation was good and no issues observed. However, I have changed the indentation from the tabs to Spaces and committed the changes. :)

williammatherjones replied:
I'm still seeing tabs in the PR.  There is a setting in eclipse to use spaces instead of tabs when formatting code.  If you have not enabled that, I suggest you change that and format the code again.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+
+	// Add/Create Authentication record in database. Create User and grant the
+	// permissions for User
+	private static void SetUpDbForOAuth() throws Exception {
```
In java convention for method names is to use camelCase. It's okay for the casing of the names in this sample to differ from the names in the ADO implementation in order to follow the convention. Please ensure all method names start with a lowercase letter and adhere to camelCase.

rvpanchal1202 replied:
Changed all method names to camelCase.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+	private static void SetUpDbForOAuth() throws Exception {
+		String connectionString = getConnectionString();
+		vConnection = DriverManager.getConnection(connectionString);
+		String USER = prop.getProperty("User");
```
As these are not  constant it should not be in all caps with underscore spacing. Please change these to be camelCase

rvpanchal1202 replied:
Changed non constant vars to camelCase.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+		st.execute("DROP USER IF EXISTS " + USER + " CASCADE;");
+		st.execute("DROP AUTHENTICATION IF EXISTS v_oauth CASCADE;");
+		st.execute("CREATE AUTHENTICATION v_oauth METHOD 'oauth' LOCAL;");
+		st.execute("ALTER AUTHENTICATION v_oauth SET client_id= '" + CLIENT_ID + "';");
```
nit: either none should have spaces before the equal sign, or they all should. I would prefer that they all do, but lets make them consistent

rvpanchal1202 replied:
Done.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+		String CLIENT_SECRET = prop.getProperty("ClientSecret");
+		String DISCOVERY_URL = prop.getProperty("DiscoveryUrl");
+		Statement st = vConnection.createStatement();
+		st.execute("DROP USER IF EXISTS " + USER + " CASCADE;");
```
Can you put all of these commands into an array of strings, then iterate through the array and call st.execute() for each element of the array. I asked Kevin to make that change in the ADO sample previously as well. It makes it easier to see everything that is being executed and it's easy to modify if need be.

rvpanchal1202 replied:
Added queries to a string array and executed the queries using iterating through the array.

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+	private static void TearDown() {
+		try {
+			Statement st = vConnection.createStatement();
+			String USER = prop.getProperty("User");
```
not a constant, change to `user`

rvpanchal1202 replied:
Done

DMickens commented on JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java:
```diff
@@ -10,91 +10,309 @@
+	// If access token is Invalid/Expired, Get new tokens using password/refresh grant
+	private static void EnsureAccessToken() throws Exception {
+		try {
+			String accessToken = System.getenv(OAUTH_ACCESS_TOKEN_VAR_STRING);
```
System.getenv will not throw an exception, just return null. So these should be moved outside of the try block

rvpanchal1202 replied:
Moved System.getenv outside of try block

DMickens commented:
I really want to be able to compare this side by side with the ado.net sample application, but the spacing issue is making that very difficult. When splitting the screen to review side by side, this code is stretched out and split over many lines with large spaces in between sections of code and it makes it difficult to review. Would you mind fixing that issue first and committing the change? Then I will finish my review, I'm not sure if it's an issue from using tabs or just too many spaces or what but making indentation level 4 manual spaces should fix it.

rvpanchal1202 commented:
Code rearrangement of methods are done with respect to ADO.Net Sample App.

rvpanchal1202 committed:
Update OAuthSampleApp.java

Review comments addressed, 
-> Code alliagnment 
-> Contant and non constant variables consistency adjusted,
-> functions are rearranged as per ADO.Net Sample app

williammatherjones reviewed (changes requested):
Other than the tab/space issue, the sample looks really good.   Please push a commit with the formatting changes and I'll approve.
